docs/architecture/design-proposals/dashboard/deployment-improvements.md

```diff
@@ -0,0 +1,197 @@
+
+1. The deployment component loads the details of the Chart. If the chart doesn't contain the file `values.schema.json`, continue as today; Only the advanced deployment is allowed. In other case, go to the next step.
+2. Load the `values.schema.json`. If it doesn't contain the `kubeapps` key, auto-generate the form using the tool `react-jsonschema-form`. From this point, two different tabs will be shown to the user. If the user changes tab from `basic` to `advanced` the new form will be shown but the values won't be ported. If the `kubeapps` key is defined go to the next step.
+3. Kubeapps will support ceratin types of parameters (like `username` or `resources`). If those parameters are present, we will render the rich view of the parameter, for example card-choice instead of a text-box. For parameters not supported, we should be able to fallback to the automatic generation. This may not be possible so in that case the parameter will be ignored (with an error for the developer console). If a parameter has a `title` or a `description`, those should be used in the form.
+4. Once the user clicks on "submit", we can use the `kubeapps` mapping to generate a JSON object. We can also use the JSON schema to validate the the processed object complies with the definition.
+
+## Upgrade Workflow
+
+The upgrade workflow will be similar to the deployment with some caveats:
+
+1. If the application to be upgraded doesn't contain a `values.schema.json`, behave as today. In case the current version contains the file, it should be included in new versions. If that's the case, go to the next step.
+2. If the schema contains the `kubeapps` key, use the current `values` to pre-populate the information of the form. If the schema doesn't contain that key, the renderization gets done by the automatic form generator so we may not be able to pre-fill the app information.
+3. Once the user selects a new version, the new `schema` gets loaded. For parameters existing in the previous version, re-fill the form with the previous values. In case there is a new parameter in the form, leave it empty (or use its default value). If a parameter is no longer used, ignore it.
+4. The logic to generate the final object and validate it is the same than for the deployment.
+
+## Development
+
+The amount of code/time required to implement this feature is considerably high. To avoid large PRs or temporary branches, we suggest to incrementaly implement this feature, merging changes on `master` but protecting the current status using a feature flag. If the feature flag is disabled (default behavior), the current deployment and upgrade form will be shown. For developers (or anyone that wants to test the feature), we will enable this flag to activate the new form. We can store this flag in the Kubeapps ConfigMap. Once the feature is finished, this flag will be removed from the configuration, enabling the feature for everyone.
```

Comment:
+100.
Seems like the steps could be planned too - just off the top of my head:
 * Add support for loading `values.schema.json` when present
 * Validate submitted values.json (behind flag)
 * component to parse values.yaml using schema to redux state
 * update to deploy values from the redux state directly, rather than the input field
 * add component to render form based on selected fields and json schema
 * update component so that form changes update the redux state for the values
 * update the upgrade to work similarly
 * profit!

(again, that's just me thinking out aloud - you might have a different idea - great) EDIT: I see you've got something similar down below :)

Reply:
Sounds good to me! :+1: